### PR TITLE
Sync from model config on agent startup

### LINKED ISSUE #1203: MMS model agent gets panic when downloading models
/kind bug

**What steps did you take and what happened:**
[A clear and concise description of what the bug is.]

```
{"level":"info","ts":1605215785.7493982,"logger":"Downloader","msg":"Downloading to model dir","modelUri":"s3://torchscript","modelDir":"/mnt/models"}
{"level":"info","ts":1605215785.7495825,"logger":"modelAgent","msg":"Download model ","modelName":"cifar10","storagUri":"s3://torchscript","modelDir":"/mnt/models"}
panic: runtime error: index out of range [1] with length 1

goroutine 22 [running]:
github.com/kubeflow/kfserving/pkg/agent/storage.(*S3Provider).DownloadModel(0xc00048d9a0, 0x7ffed6e0f493, 0xb, 0xc000580480, 0x7, 0xc0005804a0, 0x10, 0x1b, 0x0)
	github.com/kubeflow/kfserving/pkg/agent/storage/s3.go:59 +0x49c
github.com/kubeflow/kfserving/pkg/agent.(*Downloader).download(0xc000286198, 0xc000580480, 0x7, 0xc0005804a0, 0x10, 0x2, 0x3a)
	github.com/kubeflow/kfserving/pkg/agent/downloader.go:75 +0xff
github.com/kubeflow/kfserving/pkg/agent.(*Downloader).DownloadModel(0xc000286198, 0xc000580480, 0x7, 0xc00045c540, 0x0, 0x0)
	github.com/kubeflow/kfserving/pkg/agent/downloader.go:51 +0x76a
github.com/kubeflow/kfserving/pkg/agent.(*Puller).modelProcessor(0xc000286180, 0xc000580480, 0x7, 0xc00045c5a0)
	github.com/kubeflow/kfserving/pkg/agent/puller.go:131 +0x2b9
created by github.com/kubeflow/kfserving/pkg/agent.(*Puller).enqueueModelOp
```

**What did you expect to happen:**

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- Istio Version:
- Knative Version:
- KFServing Version:
- Kubeflow version:
- Kfdef:[k8s_istio/istio_dex/gcp_basic_auth/gcp_iap/aws/aws_cognito/ibm]
- Minikube version:
- Kubernetes version: (use `kubectl version`):
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by yuzisun**
**What this PR does / why we need it**:
Model agent does not sync models from config when scaling to more replicas as there is no CREATE event generated when pods are started with initial mounted config.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1203 

**Special notes for your reviewer**:
- Fix s3 uri parse issue when there is only bucket
- Force syncing the model config on startup, if we delete and recreate the pod, the agent should sync the model config and download/load the exiting models from model config.
```
{"level":"info","ts":1605267738.2322607,"logger":"modelAgent","msg":"Initializing model agent with","config-dir":"/mnt/configs","model-dir":"/mnt/models"}
{"level":"info","ts":1605267738.2323687,"logger":"modelAgent","msg":"Initializing s3 client with ","endpoint":"http://minio-service:9000","region":"us-west-2"}
{"level":"info","ts":1605267738.2324173,"logger":"Watcher","msg":"Syncing model directory..","modelDir":"/mnt/models"}
{"level":"info","ts":1605267738.2325592,"logger":"modelWatcher","msg":"adding model","modelName":"cifar"}
{"level":"info","ts":1605267738.232946,"logger":"modelProcessor","msg":"worker is started for","model":"cifar"}
{"level":"info","ts":1605267738.233302,"logger":"modelProcessor","msg":"Downloading model","storageUri":"s3://torchscript/cifar"}
{"level":"info","ts":1605267738.2334309,"logger":"Downloader","msg":"Downloading to model dir","modelUri":"s3://torchscript/cifar","modelDir":"/mnt/models"}
{"level":"info","ts":1605267738.2335713,"logger":"modelAgent","msg":"Download model ","modelName":"cifar","storagUri":"s3://torchscript/cifar","modelDir":"/mnt/models"}
{"level":"info","ts":1605267738.2337146,"logger":"Watcher","msg":"Start to watch model config event"}
{"level":"info","ts":1605267738.2337751,"logger":"Watcher","msg":"Watching","modelConfig":"/mnt/data/models.json"}
{"level":"info","ts":1605267738.626316,"logger":"modelOnComplete","msg":"completion event for model","modelName":"cifar","inFlight":0}
```

**Release note**:

```release-note

```

### CODE REVIEW

File: pkg/agent/watcher.go [hunk 1]
```diff
@@ -43,18 +44,40 @@ func NewWatcher(configDir string, modelDir string) Watcher {
 	if err != nil {
 		log.Error(err, "Failed to sync model dir")
 	}
-	return Watcher{
+	watcher := Watcher{
 		configDir:    configDir,
 		modelTracker: modelTracker,
-		ModelEvents:  make(chan ModelOp),
+		ModelEvents:  make(chan ModelOp, 100),
```
Comment (yuzliu) [inline pkg/agent/watcher.go:50]: I thought we intentionally made this a blocking channel? Can @ifilonenko can confirm?

Comment (yuzisun) [reply pkg/agent/watcher.go:50]: I do not see much reason to make it blocking channel, the problem is that if we push the events first and then start the puller the process just hangs there.

**Review by ifilonenko — APPROVED** [review]
/lgtm good catch on testing the scaling

**Pull request merged**